### PR TITLE
feat: parsing config file using AST

### PR DESCRIPTION
**Pull request by anshulsahni**
<div id='description'>
<h3>Summary by Bito</h3>
This pull request implements an AST-based approach for configuration file parsing using Peggy-generated grammar. It introduces new definitions for code owner configurations, path expressions, and ownership rules with improved error handling and location tracking. The changes enhance maintainability and extensibility of the configuration parsing system.
<br>
<br>
<b>Unit tests added</b>: False
<br>
<br>
<b>Estimated effort to review (1-5, lower is better)</b>: 3
</div>

## Summary by CodeRabbit

- **New Features**
	- Introduced support for parsing advanced code ownership configuration files, allowing path patterns, ownership rules with logical operators, and optional minimum review counts.
	- Added a parser that provides detailed error messages and outputs structured results for configuration inputs.
- **Tests**
	- Added a demonstration script to visualize parsed configuration structures for sample inputs.

### CODE REVIEW

**Comment by anshulsahni** [comment]
@coderabbitai review

File: AST/ConfigGrammer.peggy [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+CODEOWNER "codeowner"
+ = "@" chars:[a-zA-Z0-9_-]+ digits:("["[0-9]+"]")? space*
+ {
+ 	return {
+   		type: "CodeOwner",
+        value: chars.join(""),
+        minReviews: digits ? parseInt(digits[1].join(""),10) : null
+    }
+ }
```
[the remark this answers is not in the document]
Comment (anshulsahni) [reply AST/ConfigGrammer.peggy:46]: What will be the impact, if we don't do this change

[the remark this answers is not in the document]
Comment (anshulsahni) [reply AST/ConfigGrammer.peggy:46]: Are there more alternatives?

**Review by Kirat2508 — COMMENTED** [review]
# PR Review: CODEOWNER++ Grammar Configuration

Hey there! Thanks for sharing this PEG.js grammar for the CODEOWNER++ config. I can see you're working on parsing code ownership rules with a more expressive syntax than the standard CODEOWNERS format. I like the direction this is going - the logical operators will make our ownership rules much more flexible.

Here are some thoughts after reviewing the grammar:

## Overall Structure

The parser looks good overall, but I noticed a few areas we could improve:

### File Header
* **Lines 1-4**: The header comments appear to be leftover from a template:
```javascript
// Simple Arithmetics Grammar
// ==========================
//
// Accepts expressions like "2 * (3 + 4)" and computes their value.
```
These should be updated to reflect what this grammar actually does (parsing CODEOWNER++ configurations).

### PathExpression Structure
* **Lines 13-18**: The returned object structure:
```javascript
{
  return {
    path:path,ownership_rules:ownership_rule
  }
}
```

The property spacing is inconsistent. Let's format it for consistency:
```javascript
{
  return {
    path: path,
    ownershipRules: ownership_rule
  }
}
```
Also note I changed `ownership_rules` to camelCase to match our team's JavaScript/TypeScript conventions.

### OWNERSHIP_RULE Definition
* **Line 21**: The rule:
```javascript
= (CODEOWNER (OWNERSHIP_OPERATOR CODEOWNER)*)+
```

This looks valid, but the double nesting with two `+` quantifiers makes it a bit hard to reason about. Could we simplify this to something more explicit? For instance:
```javascript
= CODEOWNER (OWNERSHIP_OPERATOR CODEOWNER)*
```

If we're trying to support groups of ownership rules, we might want to make that more explicit in the grammar with named rules.

### Return Value Formatting
* **Lines 24-30**: The PATHGLOB's return object:
```javascript
{
  return {
    type: "PathGlob",
    value: chars.join("")
  }
}
```

The indentation is a mix of spaces and tabs. Let's keep it consistent throughout.

### Error Handling
There doesn't appear to be any explicit error handling for invalid syntax. While PEG.js will generate errors automatically, consider adding some custom error messages for common mistakes (like forgetting the `@` in a CODEOWNER).

## Missing Features

I notice a few things that might be valuable to add:

1. **Comments Support**: Many CODEOWNERS files use `#` for comments. We should add support for those.

2. **Parentheses for Grouping**: Complex ownership rules might benefit from parentheses to explicitly control operator precedence.

3. **Empty Lines Handling**: The grammar allows newlines but doesn't explicitly handle completely empty lines (which are common in config files for readability).

## Security Considerations

* **Lines 38-46**: The CODEOWNER parsing:
```javascript
= "@" chars:[a-zA-Z0-9_-]+ digits:("["[0-9]+"]")? space*
{
  return {
    type: "CodeOwner",
    value: chars.join(""),
    minReviews: digits ? parseInt(digits[1].join(""),10) : null
  }
}
```

This does handle the parsing of things like `@username[2]` correctly, but be cautious with direct `parseInt()` without validation. Since this is a parser, the input digits are already validated by the grammar pattern, but it's still good practice to add a bounds check to prevent extremely large numbers that could cause performance issues.

## Testing Considerations

While not part of this PR, I'd recommend we create a comprehensive test suite that covers:
- Basic path patterns
- Complex ownership rules with multiple operators
- Edge cases (empty files, malformed inputs)
- All the special characters allowed in path globs

## Overall

This is a solid foundation for the CODEOWNER++ grammar! The logical operators will be a significant improvement over the standard format. With some minor tweaks for consistency and a few enhancements for usability, this should serve us well.

Would you like to pair on adding comment support or any of the other suggested improvements? I've got some time tomorrow afternoon if you'd like to collaborate.

Cheers,
Your Tech Lead
# PR Review: Testing Parser Output Script

Hey there! Thanks for submitting this PR with a test script for the parser. I can see this is a new file that helps visualize the parser output, which is a great way to validate functionality.

## Overall Impression
This looks like a helpful debugging/testing utility to inspect the parser's output. Using `util.inspect` is a good choice for visualizing complex objects. I have a few thoughts to make this even more useful.

## Specific Feedback

### `index.js` (Lines 1-12)

1. **Lines 1-2**: The code imports the parser module directly:
   ```javascript
   const parser = require("./parser");
   const util = require("util");
   ```
   This assumes the parser module is in the same directory. If we're planning to use this across the codebase, we might want to consider using a more robust import path. Is `parser.js` definitely in the same folder as this new file?

2. **Lines 6-8**: The test input contains duplicate lines:
   ```javascript
   parser.parse(`
   path/to/*.txt @anshul && @anshulsahni && @saloni[2] || @arpit
   path/to/*.txt @anshul && @anshulsahni && @saloni[2] || @arpit
   `)
   ```
   Was this duplication intentional? If you meant to test different patterns, we should vary the input strings.

3. **Lines 4-11**: This is a console.log-based approach:
   ```javascript
   console.log(
     util.inspect(
       parser.parse(`...`),
       { showHidden: false, depth: null, colors: true }
     )
   );
   ```
   Since we're building a SaaS platform with proper testing practices, should we consider:
   
   - Adding this as a proper Jest/Mocha test that asserts on expected output?
   - Turning this into a more reusable utility function that other developers can use?
   - Adding some error handling in case parser.parse throws an exception?

4. **General**: There are no comments explaining what this file is for. A brief header comment would help other developers understand its purpose.

## Suggestions for Improvement

1. Consider adding a proper description in comments:
   ```javascript
   /**
    * Parser output visualization script
    * 
    * This utility helps visualize the output of our CODEOWNERS parser
    * to aid in debugging and development.
    * 
    * Usage: node index.js
    */
   ```

2. Make the code more robust by adding error handling:
   ```javascript
   const parser = require("./parser");
   const util = require("util");

   try {
     const result = parser.parse(`
       path/to/*.txt @anshul && @anshulsahni && @saloni[2] || @arpit
       path/to/*.js @mei && @jayden || @rohan
     `);
     
     console.log(util.inspect(result, { showHidden: false, depth: null, colors: true }));
   } catch (error) {
     console.error("Error parsing input:", error);
   }
   ```

3. Consider making this more interactive by accepting input from command line:
   ```javascript
   const parser = require("./parser");
   const util = require("util");
   const fs = require("fs");

   const inputFile = process.argv[2];
   let input;

   if (inputFile) {
     try {
       input = fs.readFileSync(inputFile, 'utf8');
     } catch (err) {
       console.error(`Could not read file ${inputFile}:`, err);
       process.exit(1);
     }
   } else {
     input = `
       path/to/*.txt @anshul && @anshulsahni && @saloni[2] || @arpit
       path/to/*.js @mei && @jayden || @rohan
     `;
     console.log("No input file provided, using default example input.");
   }

   try {
     const result = parser.parse(input);
     console.log(util.inspect(result, { showHidden: false, depth: null, colors: true }));
   } catch (error) {
     console.error("Error parsing input:", error);
   }
   ```

## Questions

1. What's the intended use case for this script? Is it meant as a one-off debugging tool, or will it be used regularly?

2. Are we planning to add unit tests for the parser itself? This visualization is helpful, but doesn't verify correctness programmatically.

Overall, this is a simple but useful addition for debugging. With a few small enhancements, it could be even more valuable to the team. Let me know if you'd like to discuss any of these suggestions!

Cheers,
Your Tech Lead
# Code Review: CODEOWNER_PLUS_PLUS Parser

Hi there! I've reviewed the Peggy parser you've contributed for our CODEOWNER_PLUS_PLUS format. The implementation looks solid overall and I can see it's handling the path globs, code owners, and logical operators as designed. Nice work on structuring the grammar rules cleanly!

Here are some thoughts and suggestions:

## General Observations

- The parser appears to be a code-generated output from Peggy 3.0.2, with custom grammar rules for parsing our CODEOWNER_PLUS_PLUS configuration format
- The implementation supports path globs, code owners with optional review counts (e.g., `@username[2]`), and logical operators (`&&`, `||`)
- Good job providing helpful error messages with location information

## Specific Feedback

### Parser Structure and Logic

**Lines 499-517: `peg$parsegrammar`**
The entry point to the parser is very simple, which is good. It just delegates to `parseCODEOWNER_PLUS_PLUS_CONFIG()`.

**Lines 731-752: `peg$parsePATHGLOB`**
```javascript
var peg$r0 = /^[A-Za-z0-9\/*_.\-]/;
// ...
function peg$parsePATHGLOB() {
  // ...
}
```
The path glob regex only allows certain characters. Consider if we should support more complex glob patterns that might include characters like `{` and `}` for matching alternatives (e.g., `src/{js,ts}/*.js`). If we need this in the future, we'd have to update the regex.

**Line 816-892: `peg$parseCODEOWNER`**
I like how you implemented the optional review count syntax (`@username[2]`). The implementation correctly parses this and converts it to a number via `parseInt`.

**Lines 270-322: Error formatting**
The error formatting looks comprehensive but there are multiple string concatenation operations:

```javascript
str +=
  "\n --> " +
  loc +
  "\n" +
  filler +
  " |\n" +
  offset_s.line +
  " | " +
  line +
  "\n" +
  filler +
  " | " +
  peg$padEnd("", s.column - 1, " ") +
  peg$padEnd("", hatLen, "^");
```

Template literals would be cleaner, but I understand this is machine-generated code, so it's not a critical issue.

### Potential Edge Cases

**Lines 261-275: Error Reporting**
The error formatting assumes `s.line === e.line` might be true or false, handling both cases. Good defensive programming!

**Lines 298-306: Character Escaping**
The character escaping in `literalEscape` is thorough, accounting for various control characters and special cases.

**Line 411-412: Location Computation with Source Offset**
```javascript
res.start = peg$source.offset(res.start);
res.end = peg$source.offset(res.end);
```
I like that there's a check for `typeof peg$source.offset === "function"` before calling it - this prevents errors if the source doesn't provide an offset function.

### Performance Considerations

**Lines 454-459: Parser Position Caching**
```javascript
function peg$computePosDetails(pos) {
  var details = peg$posDetailsCache[pos];
  // ...
}
```
The parser caches position details which helps with performance, especially for large files with many repeated structures. Good approach!

**Lines 519-583: Path Expression Parsing**
The parsing of path expressions appears to use a non-greedy approach which should perform well even on complex configurations.

### TypeScript Considerations

Since we're mainly a TypeScript shop, it would be helpful to have type definitions for the parser output. The parser is exporting these structures:

```javascript
var peg$f0 = function (path, ownership_rule) {
  return {
    path: path,
    ownership_rules: ownership_rule,
  };
};
var peg$f2 = function (chars, digits) {
  return {
    type: "CodeOwner",
    value: chars.join(""),
    minReviews: digits ? parseInt(digits[1].join(""), 10) : null,
  };
};
```

Would you be open to creating an accompanying `.d.ts` file to make it easier to consume this parser in our TypeScript codebase? Something like:

```typescript
export interface CodeOwnerConfig {
  path: PathGlob;
  ownership_rules: OwnershipRule[];
}

export interface PathGlob {
  type: "PathGlob";
  value: string;
}

export interface CodeOwner {
  type: "CodeOwner";
  value: string;
  minReviews: number | null;
}

export interface OwnershipOperator {
  type: "OwnershipOperator";
  operator: "and" | "or";
}

export type OwnershipRule = [CodeOwner, Array<[OwnershipOperator, CodeOwner]>];

export declare function parse(input: string, options?: object): CodeOwnerConfig[];
```

### Documentation

I notice there aren't any JSDoc comments. Since this is a parser that will be used by other developers on our team, some documentation about:

1. What the `parse` function expects and returns 
2. Any special options that can be passed
3. Common error scenarios and how to handle them

would be helpful. This could be in the form of JSDoc comments or a separate README.

## Summary

This is a solid implementation of the CODEOWNER_PLUS_PLUS parser. The grammar rules are well-structured and the error handling is robust. The main considerations are:

1. Consider adding TypeScript type definitions
2. Add documentation to make it easier for other team members to use
3. Evaluate if the path glob pattern should support more complex matching options

Let me know if you'd like to discuss any of these points in more detail. I'm available for a chat if you want to go through anything specific.

Great work!
